### PR TITLE
[api] fixes #1049 - Status of trusted peers exposed in API

### LINKED ISSUE #1049: Record when trusted peers cannot be reached and expose over API
See #1044 

If a trusted peer cannot be reached, we should collect this info and expose it in an API, possible in the `/health` API #1039  and/or in a `/network/` API endpoint

### PR DESCRIPTION
**Pull request by stdevHan**
Fixes #1049 

Changes:
-Add /network/status endpoint

Does this change need to mentioned in CHANGELOG.md?
yes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -60,7 +60,7 @@ else
 endif
 
 run:  ## Run the skycoin node. To add arguments, do 'make ARGS="--foo" run'.
-	go run cmd/skycoin/skycoin.go --gui-dir="./${STATIC_DIR}" --launch-browser=true ${ARGS}
+	./run.sh ${ARGS}
```
Comment (olemis) [inline Makefile]: Why is this change necessary to implement the new endpoint ?

File: ci-scripts/integration-test-stable.sh [hunk 2]
```diff
@@ -57,7 +61,7 @@ fi
 # Compile the skycoin node
 # We can't use "go run" because this creates two processes which doesn't allow us to kill it at the end
 echo "compiling skycoin"
-go build -o "$BINARY" cmd/skycoin/skycoin.go
+go build -o "$BINARY" -ldflags "${GOLDFLAGS}" cmd/skycoin/skycoin.go
```
Comment (olemis) [inline ci-scripts/integration-test-stable.sh]: Why d'u need changes in this file to implement the new endpoint ?

File: ci-scripts/integration-test-stable.sh [hunk 3]
```diff
@@ -20,6 +20,10 @@ RUN_TESTS=""
 # run wallet tests
 TEST_WALLET=""
 
+COMMIT=$(git rev-parse HEAD)
```
Comment (olemis) [inline ci-scripts/integration-test-stable.sh]: Why d'u need changes in this file to implement the new endpoint ?

File: run.sh [hunk 4]
```diff
@@ -1,7 +1,13 @@
 #!/usr/bin/env bash
 
+set -x
```
Comment (olemis) [inline run.sh]: Why d'u need changes in this file to implement the new endpoint ?

File: ci-scripts/integration-test-disable-wallet-api.sh [hunk 5]
```diff
@@ -20,6 +20,10 @@ RUN_TESTS=""
 # run wallet tests
 TEST_WALLET=""
 
++COMMIT=$(git rev-parse HEAD)
```
Comment (olemis) [inline ci-scripts/integration-test-disable-wallet-api.sh]: Why d'u need changes in this file to implement the new endpoint ?

File: cmd/skycoin/skycoin.go [hunk 6]
```diff
@@ -33,8 +33,9 @@ var (
 	Version = "0.22.0"
 	// Commit id
 	Commit = ""
-
-	help = false
+	// Branch name
```
Comment (olemis) [inline cmd/skycoin/skycoin.go]: Why d'u need changes in this file to implement the new endpoint ?

File: src/gui/network.go [hunk 7]
```diff
@@ -8,6 +8,60 @@ import (
...
+
+// ConnectionsHealth struct
+type ConnectionsHealth struct {
+	Count        int                `json:"count"`
+	TotalAlive   int                `json:total_alive`
+	TotalOffline int                `json:total_offline`
+	Connections  []ConnectionStatus `json:"connections"`
+}
+
+func defaultStatus(gateway Gatewayer) ConnectionsHealth {
```
Comment (olemis) [inline src/gui/network.go]: 👍

File: src/visor/visor.go [hunk 8]
```diff
@@ -75,6 +75,7 @@ func DropletPrecisionCheck(amount uint64) error {
 type BuildInfo struct {
 	Version string `json:"version"` // version number
 	Commit  string `json:"commit"`  // git commit id
+	Branch  string `json:"branch"`  // git branch name
```
Comment (olemis) [inline src/visor/visor.go]: Why d'u need changes in this file to implement the new endpoint ?

File: src/gui/README.md [hunk 9]
```diff
@@ -86,6 +88,57 @@ Result:
...
+Response:
+
+```json
+{
+    "count": 8,
+    "TotalAlive": 8,
+    "TotalOffline": 0,
+    "connections": [
+        {
+            "Connection": "104.237.142.206:6000"
```
Comment (olemis) [inline src/gui/README.md]: `Status` is missing in all these examples

File: src/gui/README.md [hunk 10]
```diff
@@ -86,6 +88,65 @@ Result:
...
+Response:
+
+```json
+{
+"count": 8,
+"total_alive": 8,
+"total_offline": 0,
+"Connections": [
+{
+"Status": "Connected",
```
Comment (olemis) [inline src/gui/README.md]: This should be `IsAlive` set to true/false .

File: src/gui/network.go [hunk 11]
```diff
@@ -8,6 +8,64 @@ import (
 	wh "github.com/skycoin/skycoin/src/util/http" //http,json helpers
 )
 
+// ConnectionStatus structs
+type ConnectionStatus struct {
+	Status     string `json:"Status"`
```
Comment (olemis) [inline src/gui/network.go]: Remove this field.

File: src/gui/network.go [hunk 12]
```diff
@@ -8,6 +8,64 @@ import (
...
+	Connection string `json:"Connection"`
+	isAlive    bool   `json:"isalive"`
```
Comment (olemis) [inline src/gui/network.go]: Name should start with capital `I` i.e. `IsAlive` for this field to be included in response .

File: Makefile [hunk 13]
```diff
@@ -147,4 +147,4 @@ clean: ## Clean dist files and delete all builds in electron/release
 	rm $(ELECTRON_DIR)/release/*
 
 help:
-	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
Comment (olemis) [inline Makefile]: Whitespaces should not pollute the PR diff. Make sure this file does not get listed .

File: ci-scripts/integration-test-disable-wallet-api.sh [hunk 14]
```diff
@@ -116,4 +116,4 @@ elif [[ (-z $TEST || $TEST = "cli") && $CLI_FAIL -ne 0 ]]; then
 else
   exit 0
 fi
-# exit $FAIL
+# exit $FAIL
```
Comment (olemis) [inline ci-scripts/integration-test-disable-wallet-api.sh]: Whitespaces should not pollute the PR diff. Make sure this file does not get listed .

File: ci-scripts/integration-test-stable.sh [hunk 15]
```diff
@@ -115,4 +115,4 @@ elif [[ (-z $TEST || $TEST = "cli") && $CLI_FAIL -ne 0 ]]; then
 else
   exit 0
 fi
-# exit $FAIL
+# exit $FAIL
```
Comment (olemis) [inline ci-scripts/integration-test-stable.sh]: Whitespaces should not pollute the PR diff. Make sure this file does not get listed .

File: run.sh [hunk 16]
```diff
@@ -1,7 +1,10 @@
 #!/usr/bin/env bash
 
+
```
Comment (olemis) [inline run.sh]: Whitespaces should not pollute the PR diff. Make sure this file does not get listed .

File: run.sh [hunk 17]
```diff
@@ -1,7 +1,10 @@
...
+
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 echo "skycoin binary dir:" "$DIR"
 pushd "$DIR" >/dev/null
+
```
Comment (olemis) [inline run.sh]: Whitespaces should not pollute the PR diff. Make sure this file does not get listed .

**Comment by olemis** [comment]
Except for some minor whitespace changes , I think this is ok. @gz-c could u pls review ?

File: src/gui/gatewayer_mock_test.go [hunk 18]
```diff
@@ -968,7 +967,7 @@ func (m *GatewayerMock) UpdateWalletLabel(p0 string, p1 string) error {
 func (m *GatewayerMock) GetDefaultStatus() *daemon.ConnectionsHealth {
 
 	connsDefault := m.GetDefaultConnections()
-	sort.Strings(connsDefault)
+	// sort.Strings(connsDefault)
```
Comment (gz-c) [inline src/gui/gatewayer_mock_test.go]: why is this commented

Comment (olemis) [reply src/gui/gatewayer_mock_test.go]: @stdevHan is this he line you were saying that was causing test failures depending on go lang version ? What was the version that was failing ?

File: src/daemon/gateway.go [hunk 19]
```diff
@@ -932,6 +933,21 @@ type Health struct {
 	Version            visor.BuildInfo
 	OpenConnections    int
 	Uptime             time.Duration
+	ConnectionsHealth  *ConnectionsHealth
+}
+
+// ConnectionStatus structs
+type ConnectionStatus struct {
+	Connection string `json:"connection"`
+	IsAlive    bool   `json:"isAlive"`
```
Comment (gz-c) [inline src/daemon/gateway.go]: use is_alive (underscores)

File: src/daemon/gateway.go [hunk 20]
```diff
@@ -932,6 +933,21 @@ type Health struct {
...
+}
+
+// ConnectionsHealth struct
+type ConnectionsHealth struct {
+	Count        int                `json:"count"`
+	TotalAlive   int                `json:"totalAlive"`
```
Comment (gz-c) [inline src/daemon/gateway.go]: total_alive

File: src/daemon/gateway.go [hunk 21]
```diff
@@ -932,6 +933,21 @@ type Health struct {
...
+	TotalOffline int                `json:"totalOffline"`
```
Comment (gz-c) [inline src/daemon/gateway.go]: total_offline

File: src/gui/README.md [hunk 22]
```diff
@@ -93,6 +94,64 @@ Result:
 ```
 
 ## General system checks
+### Network check
```
Comment (gz-c) [inline src/gui/README.md]: Should go in the "Network Status" section

File: src/gui/README.md [hunk 23]
```diff
@@ -93,6 +94,64 @@ Result:
...
+
+Example:
+
+```sh
+curl http://127.0.0.1:6420/network/connections/default
+```
+
+Response:
+
+```json
```
Comment (gz-c) [inline src/gui/README.md]: JSON should be formatted with indentation

File: src/gui/gatewayer_mock_test.go [hunk 24]
```diff
@@ -998,3 +999,51 @@ func (m *GatewayerMock) UpdateWalletLabel(p0 string, p1 string) error {
 	return r0
 
 }
+
+// GetDefaultStatus mocked method
+func (m *GatewayerMock) GetDefaultStatus() *daemon.ConnectionsHealth {
```
Comment (gz-c) [inline src/gui/gatewayer_mock_test.go]: The GatewayerMock methods don't need an implementation, it is generated with `goautomock` tool (you could implement it by hand, following the example of other mocked methods in this file)

File: src/gui/health.go [hunk 25]
```diff
@@ -17,10 +18,11 @@ type BlockchainMetadata struct {
...
 type HealthResponse struct {
-	BlockchainMetadata BlockchainMetadata `json:"blockchain"`
-	Version            visor.BuildInfo    `json:"version"`
-	OpenConnections    int                `json:"open_connections"`
-	Uptime             wh.Duration        `json:"uptime"`
+	BlockchainMetadata BlockchainMetadata        `json:"blockchain"`
+	Version            visor.BuildInfo           `json:"version"`
+	OpenConnections    int                       `json:"open_connections"`
+	Uptime             wh.Duration               `json:"uptime"`
+	ConnectionsHealth  *daemon.ConnectionsHealth `json:"ConnectionStatus"`
```
Comment (gz-c) [inline src/gui/health.go]: `json:"connections_status"`

File: src/gui/health.go [hunk 26]
```diff
@@ -42,15 +44,17 @@ func healthCheck(gateway Gatewayer) http.HandlerFunc {
 
 		elapsedBlockTime := time.Now().UTC().Unix() - int64(health.BlockchainMetadata.Head.Time)
 		timeSinceLastBlock := time.Second * time.Duration(elapsedBlockTime)
+		connectionshealth := gateway.GetDefaultStatus()
```
Comment (gz-c) [inline src/gui/health.go]: `connectionshealth` -> `connectionsStatus`

File: src/gui/http.go [hunk 27]
```diff
@@ -331,6 +331,7 @@ func newServerMux(c muxConfig, gateway Gatewayer, csrfStore *CSRFStore, rpc *web
 	webHandler("/network/connection", connectionHandler(gateway))
 	webHandler("/network/connections", connectionsHandler(gateway))
 	webHandler("/network/defaultConnections", defaultConnectionsHandler(gateway))
+	webHandler("/network/connections/default", defaultStatusHandler(gateway))
```
Comment (gz-c) [inline src/gui/http.go]: Change this endpoint to `/network/status`

File: src/gui/network.go [hunk 28]
```diff
@@ -83,3 +83,17 @@ func exchgConnectionsHandler(gateway Gatewayer) http.HandlerFunc {
 		wh.SendJSONOr500(logger, w, conns)
 	}
 }
+
+func defaultStatusHandler(gateway Gatewayer) http.HandlerFunc {
```
Comment (gz-c) [inline src/gui/network.go]: Rename this method to `networkStatusHandler`

File: src/gui/network.go [hunk 29]
```diff
@@ -83,3 +83,17 @@ func exchgConnectionsHandler(gateway Gatewayer) http.HandlerFunc {
...
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			wh.Error405(w)
+			return
+		}
+
+		resp := gateway.GetDefaultStatus()
+
+		wh.SendJSONOr500(logger, w, &resp)
```
Comment (gz-c) [inline src/gui/network.go]: It is not necessary to use `&resp` here, `resp` is fine.  It doesn't matter if the `resp` object is a pointer or not.

File: src/gui/network_test.go [hunk 30]
```diff
@@ -334,3 +334,105 @@ func TestGetExchgConnection(t *testing.T) {
...
+func TestGetDefaultStatus(t *testing.T) {
+
+	// ConnectionStatus structs
+	type ConnectionStatus struct {
+		Connection string
+		IsAlive    bool
+	}
+
+	// ConnectionsHealth struct
+	type ConnectionsHealth struct {
```
Comment (gz-c) [inline src/gui/network_test.go]: you can import these structs from `daemon`

File: src/gui/network_test.go [hunk 31]
```diff
@@ -334,3 +334,105 @@ func TestGetExchgConnection(t *testing.T) {
...
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+
+			endpoint := "/network/connections/default"
+
+			gateway := NewGatewayerMock()
+
+			GetDefaultConnections := []string{"11.44.66.88:9000", "44.33.22.11:9000"}
+
+			gatewayGetConnectionsResult := &daemon.Connections{
```
Comment (gz-c) [inline src/gui/network_test.go]: `gatewayGetConnectionsResult` should be a property of the test case struct

File: Makefile (hunk 13 above)
Comment (gz-c) [reply Makefile]: If you are using sublime text, add  `"ensure_newline_at_eof_on_save": true` and `"trim_trailing_white_space_on_save": true` to your preferences

File: src/gui/gatewayer_mock_test.go (hunk 24 above)
Comment (olemis) [reply src/gui/gatewayer_mock_test.go]: I applied most of the changes requested in this review , but I did not understand this one in particular . What shall be done ?

File: src/gui/README.md [hunk 32]
```diff
@@ -2199,3 +2141,61 @@ Result:
...
+{
+  "count": 8,
+    "total_alive": 8,
+    "total_offline": 0,
+    "Connections": [
```
Comment (gz-c) [inline src/gui/README.md]: Is this the correct response? These fields should not be uppercased in the response

File: src/gui/README.md [hunk 33]
```diff
@@ -2199,3 +2141,61 @@ Result:
...
+
+```sh
+curl http://127.0.0.1:6420/network/connections/default
+```
+
+Response:
+
+```json
+{
+  "count": 8,
```
Comment (gz-c) [inline src/gui/README.md]: Wrong indentation.

You can get an example of a valid response with correct formatting by making the api call to a running node

File: src/gui/README.md (hunk 23 above)
Comment (olemis) [reply src/gui/README.md]: done

File: src/gui/gatewayer_mock_test.go (hunk 24 above)
Comment (olemis) [reply src/gui/gatewayer_mock_test.go]: done !

File: CHANGELOG.md [hunk 34]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Add `password` argument to the CLI commands of `addPrivateKey`, `createRawTransaction`, `generateAddresses`, `generateWallet`, `send`
 - Support for decoding map values in cipher binary encoder
 - Add `/network/status` endpoint
+- Add `IsPermutaion` assertion to check for containers with same items but maybe in different order
```
Comment (gz-c) [inline CHANGELOG.md]: This does not belong in changelog

File: src/gui/network_test.go [hunk 35]
```diff
@@ -425,7 +426,10 @@ func TestGetDefaultStatus(t *testing.T) {
 				var msg daemon.ConnectionsHealth
 				err = json.Unmarshal(rr.Body.Bytes(), &msg)
 				require.NoError(t, err)
-				require.Equal(t, resp, msg)
+				require.Equal(t, resp.Count, msg.Count)
+				require.Equal(t, resp.TotalOffline, msg.TotalOffline)
+				require.Equal(t, resp.TotalAlive, msg.TotalAlive)
+				_require.IsPermutation(t, resp.Connections, msg.Connections)
```
Comment (gz-c) [inline src/gui/network_test.go]: sort the response in the api handler instead

File: src/gui/README.md [hunk 36]
```diff
@@ -2161,41 +2161,41 @@ Response:
...
-        {
-            "connection": "139.162.7.132:6000",
-            "is_alive": true
-        },
-        {
-            "connection": "172.104.85.6:6000",
-            "is_alive": true
-        }
-    ]
+  "total_alive": 7,
```
Comment (gz-c) [inline src/gui/README.md]: the indentation is still off

File: src/gui/README.md [hunk 37]
```diff
@@ -2161,41 +2161,41 @@ Response:
...
+  "total_offline": 1,
+  "connections": [
+    {
+      "connection": "104.237.142.206:6000",
```
Comment (gz-c) [inline src/gui/README.md]: field name should be "address"

File: CHANGELOG.md [hunk 38]
```diff
@@ -17,7 +17,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Add CLI `showSeed` command
 - Add `password` argument to the CLI commands of `addPrivateKey`, `createRawTransaction`, `generateAddresses`, `generateWallet`, `send`
 - Support for decoding map values in cipher binary encoder
-
+- Add `/network/status` endpoint
+- Add `IsPermutation` assertion to check for container types with same items but maybe in different order
```
Comment (gz-c) [inline CHANGELOG.md]: Remove this from changelog

File: src/gui/README.md (hunk 23 above)
Comment (gz-c) [reply src/gui/README.md]: Needs to be the same format as the other examples here

File: CHANGELOG.md [hunk 39]
```diff
@@ -17,8 +17,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Add CLI `showSeed` command
 - Add `password` argument to the CLI commands of `addPrivateKey`, `createRawTransaction`, `generateAddresses`, `generateWallet`, `send`
 - Support for decoding map values in cipher binary encoder
-- Add `/network/status` endpoint
-- Add `IsPermutation` assertion to check for container types with same items but maybe in different order
+- `-verify-db` option (default true), will verify the database integrity during startup and exit if a problem is found
+- `-reset-corrupt-db` option (default false) will verify the database integrity during startup and reset the db if a problem is found
+>>>>>>> f852421312ccc1181991461ce74d762117fbb855
```
Comment (gz-c) [inline CHANGELOG.md]: bad merge
